Upgrade jenkins test harness version if not compatible

CodexRaunak opened the pull request:
Fix #554 

Add Recipe for upgrading test harness version. We are maintaining a `Map` of compatible version between the two, if the version in the pom.xml is >= it , then we don't upgrade, else upgrade the test harness. The map can be extended further to store more values.

If any of the version (jenkins version and test harness) is not referenced via the properties it gets removed from the properties via calling the RemoveProperty recipe internally. 

### Testing done

Tested the unit cases.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

CodexRaunak commented:
please review the PR so that I can work on it further if required.

jonesbusy commented:
Hi,

Thanks I will take a look at it.

In the meanwhile can we

- Include the recipe on the recipes.yml (Maybe it's worth to be a top recipe? )
- Include the recipe on the (UpgradeToRecommendCoreVersion, UpgradeToLatestJava11CoreVersion, UpgradeToLatestJava8CoreVersion, UpgradeNextMajorParentVersion)

My other concern is to have a compatibility matrix on a private method. Where did you get this matrix? On the release notes?

I'm wondering it this matrix should be extracted somewhere else. Maybe it just need to be a a parameter of the recipe?

For example

```yaml
type: specs.openrewrite.org/v1beta/recipe
name: io.jenkins.tools.pluginmodernizer.UpgradeToLatestJava8CoreVersion
displayName: Upgrade to latest LTS core version supporting Java 8
description: Upgrade to latest LTS core version supporting Java 8.
tags: ['developer']
recipeList:
...
- io.jenkins.tools.pluginmodernizer.core.UpgradeJenkinsTestHarnessVersion:
      version: <lastversion compatible with JDK 8>
...

Thanks

CodexRaunak commented:
> In the meanwhile can we
> 
> * Include the recipe on the recipes.yml (Maybe it's worth to be a top level recipe? )
> * Include the recipe on the (UpgradeToRecommendCoreVersion, UpgradeToLatestJava11CoreVersion, UpgradeToLatestJava8CoreVersion, UpgradeNextMajorParentVersion)
> 

Yes we can add it in the top level recipe, also i didn't knew that we can combine existing recipes in rewrite.yml like this,
so I used the `RemoveProperty` recipe in `UpgradeJenkinsTestHarness`  internally, which needs to be changed.

CodexRaunak commented:
I need a little clarity @jonesbusy.
Am I thinking correctly
There should be 2 cases. 
Case 1:  When the property is referenced and used in the pom.xml then we are just upgrading the property.
Case 2: When the property is not referenced, so we should upgrade the dependency version itself than the property.

Apart from it I saw this recipe `UpgradeJenkinsVersion` , which is also just upgrading the property, but what if the property is not used in pom.xml anywhere, so I think we should also consider this case of upgrading the dependency directly. This logic is managed by `UpgradeVersionProperty` recipe.

So I am thinking I should modify its logic to consider case 2 as well.
What do you think?
![image](https://github.com/user-attachments/assets/72f9d8ef-c1eb-4cb7-a55a-6f2f3cd57ae7)

jonesbusy commented:
Some of the property like the jenkins.version are used in the parent pom to get the correct dependency. So it's normal to not have any reference on the pom.xml.

For Jenkins test-harness we should upgrade to the the latest compatible. We should easily find this information on the release notes. So I would say we target only the latest support for JDK 11. Then it's the latest release until we have a latest support for JDK 17 (maybe in one year or so).

Perhaps we can have this information directly on the JDK.java enum which mapping of the latest compatible version   per JDK.

If the dependency is added but not using any property, we should upgrade it (we can probably reuse a visitor on a doAfterVisit).

What do you think about this approach?

CodexRaunak commented:
Yes, It makes sense to upgrade the last version compatible with JDK and store it in enum. JDK 11 is the stable LTS version and in future as the JDK 17 we will get the latest of it as well. Cool!

jonesbusy commented:
See https://github.com/jenkinsci/plugin-pom/pull/1082

Meaning that test-harness version must always defined by this property and neved with a direct version on the current pom.

So I would say

- We upgrade the property
- We remove any referenced dependency for `jenkins-test-harness`

CodexRaunak commented:
![image](https://github.com/user-attachments/assets/7e77037a-a1e6-4fb7-abf5-bf5a00b26343)
When I am adding the` jenkins-test-harness` property in the `upgradeToRecommendCoreVersionTest` of  `DeclarativeRecipeTest`,
Its getting removed automatically, is it beacause its using the BOM so its removing extra properties?
How do I modify test cases for this recipe?

CodexRaunak commented:
@jonesbusy is it fine? I just need to add the versions to `JDK enum`, and some other touch ups.
This recipe will upgrade the `property` to the required version, and remove any `dependency version` of `jenkins-test-harness` if any.
Also it will add the property if it doesn't exist.
All this is done via visitors provided by open rewrite.

I haven't added it as a top level recipe as its just upgrading the test harness version, so I think maybe its not worth it.
Similar case like of `UpgradeJenkinsVersion`.

jonesbusy commented:
Thanks, this look well implemented but perhaps I need @gounthar second eyes.

Removing the dependency is ok 

Upgrading is ok (But I have huge doubt with the version. It looks weird we cannot upgrade more for Jenkins 2.454 for example. I would expect an higher version.

Adding the property if not exist should be removed I think. We assume the plugin just just use the version from the parent.

Fine for me to move this information on the JDK enum.

Thanks!

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/UpgradeJenkinsTestHarnessVersion.java:
```diff
@@ -0,0 +1,84 @@
+                if (!propertyExists) {
+                    LOG.info("Adding property: " + propertyKey + " with value: " + propertyValue);
+
+                    document = (Xml.Document) new AddProperty(propertyKey, propertyValue, null, null)
```
Almost sure it must be removed. If the property is absent we don't touch it. Only removing existing dependency if they are added without using the property

gounthar replied:
Definitely! 👍

CodexRaunak commented:
> Upgrading is ok (But I have huge doubt with the version. It looks weird we cannot upgrade more for Jenkins 2.454 for example. I would expect an higher version than 2.42

yes we can go above 2.72, as you said we need to upgrade it to the latest compatible with JDK_11, I wasn't able to find the latest with it on release notes. Is it latest version still compatible with it? I was going to update it.

gounthar commented:
Here is the compatibility table I've extracted from the release notes:

Compatibility Table:
| Jenkins Test Harness Version | Jenkins WAR Version |
|-------------------------------|---------------------|
| 1609.vb2253104420b | 1.64 |
| 1626.v46b0925e70db | 1.65 |
| 1640.vf14ba9a1eb71 | 1.67 |
| 1690.vfcb_959006c07 | 1.70 |
| 1705.va_296493db_feb_ | 1.71 |
| 1721.v385389722736 | 1.72 |
| 1736.vc72c458c5103 | 1.73 |
| 1751.v11e83a_ecfe14 | 1.74 |
| 1784.v83a_b_3ce99392 | 1.76 |
| 1790.v5e753edcc48d | 1.77 |
| 1802.v9de0d87365d2 | 1.80 |
| 1816.v8138d8056949 | 1.81 |
| 1827.v6b_cc1a_74b_ed5 | 1.82 |
| 1832.v0a_c4a_fc6c2c8 | 1.86 |
| 1837.vb_6efb_1790942 | 1.87 |
| 1847.v10c7a_4c2e344 | 1.88 |
| 1868.v03b_7c6673e2e | 1.89 |
| 1873.va_ed9d52f2b_42 | 1.90 |
| 1900.v9e128c991ef4 | 1.92 |
| 1922.v91779ee45b_5e | 1.93 |
| 1934.v90a_c07cf5b_21 | 1.94 |
| 1938.ve4c94459431d | 1.95 |
| 1952.v3a_b_0cd3f5a_03 | 1.96 |
| 1964.v99f5c1337633 | 1.98 |
| 2002.v0b_78b_a_d69e5d | 1.99 |
| 2034.v41c9cb_349299 | 1.100 |
| 2042.v787a_641a_9b_26 | 1.101 |
| 2053.v0ea_6fc5d99b_f | 2.417 |
| 2058.va_7b_41a_286207 | 2.418 |
| 2062.v3efc79721e45 | 2.420 |
| 2064.vcd3b_b_8f3f2b_a_ | 2.421 |
| 2085.va_c531db_287b_d | 2.422 |
| 2090.v6d1706e434a_b_ | 2.426 |
| 2096.vcfd8fa_a_67f86 | 2.427 |
| 2104.v8d1866dd0ea_f | 2.430 |
| 2109.v930e1e518c15 | 2.432 |
| 2112.ve584e0edc63b_ | 2.433 |
| 2115.v0d2b_4a_a_e80f8 | 2.434 |
| 2129.v09f309d2339c | 2.435 |
| 2135.v8f2e9795cf3f | 2.436 |
| 2140.ve736dc2b_b_d2c | 2.437 |
| 2146.v6b_f8b_1cb_d12d | 2.438 |
| 2148.v80482520b_d2c | 2.440 |
| 2151.vf4793e4dec9a_ | 2.441 |
| 2154.v9367950e9176 | 2.442 |
| 2158.v9eed8a_eb_c04b_ | 2.443 |
| 2161.v114f3cb_78326 | 2.444 |
| 2165.vf0266a_2d187f | 2.445 |
| 2171.v048c97409d12 | 2.446 |
| 2174.v111a_b_471784f | 2.449 |
| 2182.v0138ccb_c0b_cb_ | 2.450 |
| 2185.v796448319b_7c | 2.452 |
| 2189.v1c38d94b_43b_a_ | 2.453 |
| 2193.v71b_f09ec6d46 | 2.454 |
| 2199.v5c2ef8ce998f | 2.456 |
| 2205.v7e19cc98f9c7 | 2.458 |
| 2211.v29b_b_b_2e79c5a_ | 2.460 |
| 2216.v8f497401120d | 2.461 |
| 2228.v919b_5883c444 | 2.462 |
| 2232.vf1b_f732e53d4 | 2.464 |
| 2237.v7b_1605236620 | 2.466 |
| 2241.v44c8128f7cc1 | 2.467 |
| 2248.v5fb_b_936a_1d7a_ | 2.468 |
| 2250.v03a_1295b_0a_30 | 2.470 |
| 2254.vcff7a_d4969e5 | 2.471 |
| 2265.v3da_49c8134d6 | 2.473 |
| 2270.v87a_0ea_b_54da_0 | 2.475 |
| 2272.v5db_817b_a_6f0d | 2.477 |
| 2299.v7e8d22e797d3 | 2.478 |
| 2307.v10e5d0701b_e5 | 2.479 |
| 2341.v35346d95d2b_7 | 2.480 |
| 2345.v699712948764 | 2.481 |
| 2351.v862801ec05f9 | 2.482 |
| 2353.ve3f890c6eea_f | 2.485 |
| 2355.v13b_7b_3db_04c0 | 2.486 |
| 2363.vf95c19b_2ca_72 | 2.487 |
| 2369.vee6dc8db_3c95 | 2.488 |
| 2375.v46b_da_4f6c042 | 2.491 |
| 2380.vc51ea_a_cf2711 | 2.493 |
| 2384.va_d3a_58a_4b_df5 | 2.494 |
| jenkins-test-harness-2.58 | 1.53 |
| jenkins-test-harness-2.61 | 1.54 |
| jenkins-test-harness-2.64 | 1.55 |
| jenkins-test-harness-2.65 | 1.56 |
| jenkins-test-harness-2.69 | 1.58 |
| jenkins-test-harness-2.70 | 1.59 |
| jenkins-test-harness-2.72 | 1.61 |

jonesbusy commented:
> > Upgrading is ok (But I have huge doubt with the version. It looks weird we cannot upgrade more for Jenkins 2.454 for example. I would expect an higher version than 2.42
> 
> yes we can go above 2.72, as you said we need to upgrade it to the latest compatible with JDK_11, I wasn't able to find the latest with it on release notes. Is it latest version still compatible with it? I was going to update it.

This  PR need Java 17 https://github.com/jenkinsci/jenkins-test-harness/pull/762 since it bump Jenkins version to 2.463 (weekly) and LTS 2.479.1.

[2228.v919b_5883c444](https://github.com/jenkinsci/jenkins-test-harness/releases/tag/2228.v919b_5883c444) require Java 17

So meaning 2225.v04fa_3929c9b_5 is the last one to support Java 11

I've confirmed that I can compile 2225.v04fa_3929c9b_5 with Java 11

2228.v919b_5883c444 fail with Java 11 

![Screenshot from 2025-01-26 18-04-02](https://github.com/user-attachments/assets/e4fab623-86b9-4811-b0f6-df5f96cb114d)

Also I confirm that 2.71 compile with Java 8 and 2.72 fail with Java 8. So 2.71 is the last one supporting Java 8

Can you adapt the value?

Thanks

jonesbusy commented on plugin-modernizer-core/src/main/resources/META-INF/rewrite/recipes.yml:
```diff
@@ -135,6 +135,16 @@ recipeList:
       newVersion: 4.X
   - io.jenkins.tools.pluginmodernizer.RemoveDependencyVersionOverride
 ---
+#---
```
Correct this is not needed as top level. Comments can be removed

jonesbusy commented on plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/DeclarativeRecipesTest.java:
```diff
@@ -612,8 +612,9 @@ void shouldRemoveExtraProperties() {
-                        "io.jenkins.tools.pluginmodernizer.UpgradeToRecommendCoreVersion"),
+                                "/META-INF/rewrite/recipes.yml",
+                                "io.jenkins.tools.pluginmodernizer.UpgradeToRecommendCoreVersion")
+                        .expectedCyclesThatMakeChanges(2),
```
Nope. This is not expected. One cycle must be enough.

This indicate a bug in the recipe since it's not stable after 1 cycle

jonesbusy replied:
I saw some occurrence of multiple cycle on recipe order. Sometimes it's about direct visitor invocation vs doAfterVisit.

Sometimes it's about direct XML manipulation vs Maven manipulation (maybeRefreshModel?)

Only solution is by checking logs and taking debugger to see the difference between first invocation and second.

CodexRaunak replied:
> I saw some occurrence of multiple cycle on recipe order. Sometimes it's about direct visitor invocation vs doAfterVisit.
> 
> Sometimes it's about direct XML manipulation vs Maven manipulation (maybeRefreshModel?)
> 
> Only solution is by checking logs and taking debugger to see the difference between first invocation and second.

I changed the Visitor to `ChangePropertyValue` which fixed the cycle problem.

gounthar commented:
Are we interested in the fact 2.356 is the last release to support Java 8?

jonesbusy commented:
For the upgrade to latest Java 8 recipe yes. We need to bump to latest supported test harness (I think it was forgotten on this PR)

gounthar commented:
That would be [2042.v787a_641a_9b_26](https://github.com/jenkinsci/jenkins-test-harness/releases/tag/2042.v787a_641a_9b_26) then.

CodexRaunak commented:
> Also I confirm that 2.71 compile with Java 8 and 2.72 fail with Java 8. So 2.71 is the last one supporting Java 8

Isn't it `2.71` or is it `2042.v787a_641a_9b_26?`

CodexRaunak commented:
I am almost done with the changes, just need to make sure that the latest for java 8 is this 2.71 or is it 2042.v787a_641a_9b_26?

gounthar commented:
As for the release logs, it is 2042.v787a_641a_9b_26.
It was still using Jenkins 1.103, which is definitely JDK8-compatible.
Next one was using a Jenkins version which was needing JDK11.
What makes you think we should use 2.71?

On Sun, Jan 26, 2025 at 10:15 PM Raunak Madan ***@***.***>
wrote:

> I am almost done with the changes, just need to make sure that the latest
> for java 8 is this 2.71 or is it 2042.v787a_641a_9b_26?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkins-infra/plugin-modernizer-tool/pull/673#issuecomment-2614594047>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4OWLNWX2CLGNHQIEI7PL2MVF55AVCNFSM6AAAAABVPFSRQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJUGU4TIMBUG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Bruno Verachten

CodexRaunak commented:
> As for the release logs, it is 2042.v787a_641a_9b_26. It was still using Jenkins 1.103, which is definitely JDK8-compatible. Next one was using a Jenkins version which was needing JDK11. What makes you think we should use 2.71?

As @jonesbusy mentioned that 2.72 is not compiling with java 8 and 2.71 is compiling, that's why I thought it, did I get something wrong?

CodexRaunak commented:
I edited the `RemoveExtraMavenProperties` recipe in recipe.yml and removed the check for `jenkins-test-harness version` or it was removing the property.
Also Now all the recipes is taking 1 cycle only.

gounthar commented:
Let's try recompiling 2042.v787a_641a_9b_26 with JDK8 and see...

On Sun, Jan 26, 2025 at 10:39 PM Raunak Madan ***@***.***>
wrote:

> As for the release logs, it is 2042.v787a_641a_9b_26. It was still using
> Jenkins 1.103, which is definitely JDK8-compatible. Next one was using a
> Jenkins version which was needing JDK11. What makes you think we should use
> 2.71?
>
> As @jonesbusy <https://github.com/jonesbusy> mentioned that 2.72 is not
> compiling with java 8 and 2.71 is compiling, that's why I thought it, did I
> get something wrong?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkins-infra/plugin-modernizer-tool/pull/673#issuecomment-2614601671>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4OWMXCKZR6CJGX4AOU3D2MVI2VAVCNFSM6AAAAABVPFSRQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJUGYYDCNRXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Bruno Verachten

jonesbusy commented:
**2042.v787a_641a_9b_26**

Java 8

![Screenshot from 2025-01-27 06-12-28](https://github.com/user-attachments/assets/ca4c9099-7169-4353-93ee-d1b6eb24d1e6)

Java 11

![Screenshot from 2025-01-27 06-13-11](https://github.com/user-attachments/assets/e73614f6-6d48-400a-b078-9082d7d82918)

According to https://www.jenkins.io/doc/book/platform-information/support-policy-java/

2.357 is the first weekly requiring Java 11. So the 2.361 cannot support Java 8

If I try an older release (for example 1999.v093b_8a_5013e1)

I have a proper enforcer message

![Screenshot from 2025-01-27 06-31-11](https://github.com/user-attachments/assets/e1297ead-c35f-4743-87fe-db5767a01b77)

It looks this release https://github.com/jenkinsci/jenkins-test-harness/releases/tag/1903.vf505ecb_63589 upgrade to Java 11

I confirm that I can compile 1900.v9e128c991ef4 with Java 8

![image](https://github.com/user-attachments/assets/93c39899-5e57-4001-82ff-87634b1065ff)

I confirm that I cannot compile 1903.vf505ecb_63589 wth Java 8

![Screenshot from 2025-01-27 06-35-29](https://github.com/user-attachments/assets/23a7c4eb-d8a8-434a-add7-c4d35695117b)

So I mean 1900.v9e128c991ef4 is the last one supporting Java 8 (Match with release notes)

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/model/JDK.java:
```diff
@@ -24,10 +24,10 @@ public enum JDK {
-    JAVA_11(11, true, "2.164.1", "2.462.3"),
-    JAVA_17(17, true, "2.346.1", null),
-    JAVA_21(21, true, "2.426.1", null);
+    JAVA_8(8, true, null, "2.346.1", "2042.v787a_641a_9b_26"),
```
```suggestion
    JAVA_8(8, true, null, "2.346.1", "1900.v9e128c991ef4"),
```

See my latest comment

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/UpgradeJenkinsTestHarnessVersion.java:
```diff
@@ -0,0 +1,66 @@
+     * The latest compatible version.
+     */
+    @Option(displayName = "Version", description = "The version.", example = "2.70")
+    String latestCompatibleVersion;
```
If we have this information on the JDK can we get rid of this option?

If we use somethink like

```java
PluginMetadata pomMetadata = new PomResolutionVisitor().reduce(tree, new PluginMetadata());

// Can be an utils on JDK? That take a Jenkins version and return directly the latest test-harness compatible?
String latestTestHarnessVersion = JDK.min(new HashSet<>(JDK.get(pomMetadata.getJenkinsVersion())));
```

CodexRaunak commented:
I added the utility function in `JDK enum` which takes jenkins-core-version as params and give me the latest test harness version
for the oldestCompatibleJDK. Taking `oldestCompatibleJDK` also ensures stability.

Added test case for it as well.
Also by default if the jenkins version for example is `2.479.1` for which the latest test harness doesn't exist , it return the latest supported by Java 11. We will add it in future.

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/model/JDK.java:
```diff
@@ -77,6 +83,30 @@ public String getCompatibleSince() {
+
+        // Return the latest test-harness version for the oldest compatible JDK
+        return oldestCompatibleJdk != null
+                ? oldestCompatibleJdk.getLatestTestHarnessVersion() == null
```
I would keep returning null. Then the caller is responsible to do what's needed. For example use latest.release on openrewrite.

jonesbusy commented on plugin-modernizer-core/src/main/resources/versions.properties:
```diff
@@ -1,5 +1,5 @@
 openrewrite.maven.plugin.version = ${openrewrite.maven.plugin.version}
-jenkins.parent.version = 5.5
+jenkins.parent.version = 5.6
```
Thanks! Updatecli didn't run yet to update the dependency

jonesbusy commented:
Thanks. One last minor comments and we are good to merge

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/model/JDK.java:
```diff
@@ -49,16 +46,25 @@ public enum JDK {
+     * @param latestTestHarnessVersion The latest compatible version of jenkins-test-harness
      */
-    JDK(int major, boolean lts, String compatibleSince, String maximumCoreVersion) {
+    JDK(int major, boolean lts, String compatibleSince, String maximumCoreVersion, String latestTestHarnessVersion) {
```
Note for myself (out of context of this PR). This is worth adding also the parent. For example 4.88 is Java 11 and 5.x is Java 17.

It would avoid harcoding the value in several places (versions.properties or recipes) and have it centralised here.

There is some strict relation with Jenkins version and supported parent pom.xml

CodexRaunak commented:
I found a bug in the recipe related to resolving the `test harness version` based on the `Jenkins `version. When upgrading `Jenkins `to a newer version` (e.g., 2.479)`, the test harness initially resolves using the `old Jenkins version (e.g., 2.440.3),` even though the recipe execution order is correct. As a result, the test harness first upgrades according to the outdated version `(2.440.3)`. Then, after `Jenkins `is upgraded to the correct version `(2.479)`, the test harness upgrades again to align with it.

This issue causes two execution cycles in some test cases (though not all).

To fix this, we need to ensure that the test harness upgrades based on the correct Jenkins version in a single pass. 
One possible solution is to pass the updated Jenkins version as a parameter explicitly or parse the recipe.yml file to extract the target Jenkins version before running the test harness upgrade. Using tools like SnakeYAML, we can retrieve the version defined in the recipe and ensure the correct test harness version is selected upfront.


Somehow or the other we need the updated value of jenkins in the test harness. 
what do you think?

jonesbusy commented:
I see. Thanks for the detailed explaination.

I don't think parsing is a good idea, it add complexity

One other solution is to use context messaging https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices to pass state between recipes. 

It's not recommanded because of distribution of recipes. But are are not distributing recipes.

But last time I was using it I got some message from OpenRewrite it was not possible anymore.

The easiest for now is to pass the version to the recipe.

There are an other issue https://github.com/jenkins-infra/plugin-modernizer-tool/issues/688 to track filtering on recipe, so I'm not worried about maintaining the jenkins version is several places

Thanks again

jonesbusy commented:
Or an other solution is not to provide a recipe but just call the visitor on the jenkins version update recipe?

jonesbusy commented:
Then we can use cursor messaging if we need on the same recipe

CodexRaunak commented:
![image](https://github.com/user-attachments/assets/b57f434f-d953-4c2a-bd66-fe2e4a93ed12)
I got this error when using the `execution context` to pass the state between recipes, So maybe we can just pass a parameter of `jenkins version` for now, This will keep the `upgrading jenkins version` and `test harness version` recipe separated, think we should keep it isolated. 
And we can implement a scanner recipe later on.

jonesbusy commented:
> ![image](https://private-user-images.githubusercontent.com/143412156/407845864-b57f434f-d953-4c2a-bd66-fe2e4a93ed12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxNzQ4MjYsIm5iZiI6MTczODE3NDUyNiwicGF0aCI6Ii8xNDM0MTIxNTYvNDA3ODQ1ODY0LWI1N2Y0MzRmLWQ5NTMtNGMyYS1iZDY2LWZlMmU0YTkzZWQxMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOVQxODE1MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZmU2YzNiNTc5YzM1ZTc4MjNkY2EzMWQ1ZWU4YjJmN2U2NmMzZGYzNWU2MWU2MDJmNzgxNzdiYTE3NzIwMzMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9moC6cInH7eHVkm6TDCMb9rWYPd5lKAxJkjhKgGzU0M) I got this error when using the `execution context` to pass the state between recipes, So maybe we can just pass a parameter of `jenkins version` for now, This will keep the `upgrading jenkins version` and `test harness version` recipe separated, think we should keep it isolated. And we can implement a scanner recipe later on.

Yes I had the same. So it's seems it's not anymore "not recommanded" but disabled now

Yes we can just pass the "jenkinsVersion" parameter for now and improve later

CodexRaunak commented:
Now the logic depends on the caller, if not found the test harness version i.e null , it will resolve to `RELEASE` which is stable recent release.

Thanks! for the help @jonesbusy and @gounthar, I enjoy working with you.

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/UpgradeJenkinsTestHarnessVersion.java:
```diff
@@ -0,0 +1,84 @@
+                    LOG.info(
+                            "No compatible Jenkins Test Harness version found for Jenkins version {},  Defaulting to RELEASE'.",
+                            jenkinsVersion);
+                    latestCompatibleVersion = "RELEASE";
```
Shouldn't it be 'latest.release' 

​​https://docs.openrewrite.org/reference/dependency-version-selectors#supported-dependency-version-selectors

jonesbusy commented on plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/DeclarativeRecipesTest.java:
```diff
@@ -1396,6 +1463,7 @@ void upgradeNextMajorParentVersionTest() {
                                 <connection>scm:git:https://github.com/jenkinsci/empty-plugin.git</connection>
                               </scm>
                               <properties>
+                                <jenkins-test-harness.version>RELEASE</jenkins-test-harness.version>
```
https://docs.openrewrite.org/reference/dependency-version-selectors#supported-dependency-version-selectors

By using latest.release you will get the latest version

jonesbusy commented:
All good but the RELEASE is incorrect. We never use this on the Jenkins Project.

By setting latest.version you will get the latest version available.

https://docs.openrewrite.org/reference/dependency-version-selectors#supported-dependency-version-selectors

CodexRaunak commented:
`   </dependencyManagement>
     <dependencies>
     <dependency>
      <!--~~(org.jenkins-ci.main:jenkins-test-harness:latest.release failed. Unable to download POM: org.jenkins-ci.main:jenkins-test-harness:latest.release. Tried repositories:
https://repo.jenkins-ci.org/public/: HTTP 404
https://repo.maven.apache.org/maven2: HTTP 404)~~>--><dependency>
         <groupId>org.jenkins-ci.main</groupId>
         <artifactId>jenkins-test-harness</artifactId>
       </dependency>
] `
![image](https://github.com/user-attachments/assets/80589e7b-71e9-4594-86d0-df89c2920acb)
Actaully i am getting this error 404, when I set latest.release.

jonesbusy commented:
Mhh normally this is a special value to select the latest available version. But perhaps only work with some visitors.

Or what we need to do is add the latest version of test-harness on versions.property and let updatecli update it regularly

We do similar approach for the API plugin for example.

If you want to can add the latest known version on versions.properties and we can do the updatecli manifest later on

sridamul commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/model/JDK.java:
```diff
@@ -5,10 +5,7 @@
-import java.util.Comparator;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Can we remove the usage of star imports?

CodexRaunak replied:
yeah sure , my bad!

sridamul reviewed (approved):
Great work. Thanks!

jonesbusy commented:
Thanks for this huge contribution!
